### ISSUE TITLE
Can't delete join table record for Web using GraphQL

### ISSUE THREAD

**Issue by adplant** [aws-amplify/amplify-flutter#3539]
### Description

On web, I'm unable to delete a record from the auto generated table that happens when you have many to many fields. Doing the delete operation results in the following GraphQL error message:

```
GraphQLResponse<TodoTodoGroup$> error: [
  {
    "message": "Conflict resolver rejects mutation.",
    "locations": [
      {
        "line": 1,
        "column": 120
      }
    ],
    "path": [
      "deleteTodoTodoGroup"
    ],
    "errorType": "ConflictUnhandled"
  }
]
```

### Categories

- [ ] Analytics
- [ ] API (REST)
- [X] API (GraphQL)
- [ ] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Notifications (Push)
- [ ] Storage

### Steps to Reproduce

1. Have an existing record in a many to many auto generated table
2. Perform the delete operation as suggested in the docs, here (I did the `deleteById` method): https://docs.amplify.aws/lib/graphqlapi/mutate-data/q/platform/flutter/#run-a-mutation
3. Get the error

### Platforms

- [ ] iOS
- [ ] Android
- [X] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.10.1

### Amplify Flutter Version

1.2

### Deployment Method

Amplify CLI

**Comment by adplant** [comment]
I'm not sure. How do I find out?

Datastore is enabled on this backend as well.


Sincerely,

Adam Taufmann
Software Engineer
ePlant.bio
Email: ***@***.***
Phone: 808-461-7111
Time zone: Hawaii (HST)

**Comment by adplant** [comment]
Looks like we do...


In transform.conf.json we have the following, which I believe indicates
conflict resolution is enabled:

 "ResolverConfig": {
         "project": {
             "ConflictHandler": "AUTOMERGE",
             "ConflictDetection": "VERSION"
         }

**Comment by adplant** [comment]
Any ideas @ragingsquirrel3? This is definitely an immediate problem that we're trying to get fixed.

**Comment by ragingsquirrel3** [comment]
Yes, this is expected when trying to mutate a conflict resolution enabled backend without including the _version variable. If you are able to recreate the backend without conflict resolution, that is the easiest solution (assumed you don't need Datastore). We have an open feature request to improve this experience https://github.com/aws-amplify/amplify-flutter/issues/1945 which has links to many similar issues.

If you want to mutate a backend with conflict resolution using GraphQL API, you will not be able to use model mutation helpers like `deleteById` because the output requests do not include the `_version` variable in the input. You will have to make a manual string request and populate the input variable with the most recent `_version` number (which you would have to make a manual request to get from the backend in the first place).

**Comment by ragingsquirrel3** [comment]
More info on that workaround from an older issue: https://github.com/aws-amplify/amplify-flutter/issues/1434#issuecomment-1062123160.

**Comment by adplant** [comment]
Ok, thanks a lot @ragingsquirrel3 for the info. I'll confirm this works and then close the issue.

We are in a situation where we can't recreate the backend.

When looking at the example you sent the link to, it doesn't look like the version number get incremented (at least not manually). Is that true?

Also, does that example work the same way for a 'delete'?

**Comment by adplant** [comment]
@ragingsquirrel3 Well, it hasn't worked. Here's my code. Any ideas?


```
Future<String?> getPlantPlantGroupWithVersionNumber(
    {required String modelId}) async {
  const getPlantPlantGroup = 'getPlantPlantGroup';
  // request a model with the _version attached
  String gqlQueryWithVersion = '''query MyQuery {
  $getPlantPlantGroup(id: $modelId) {
    id
    _version
    plantGroupId
    plantId
  }
}''';
  final getRequest = GraphQLRequest<String>(
    document: gqlQueryWithVersion,
    decodePath: getPlantPlantGroup,
    modelType: PlantPlantGroup.classType,
  );
  final getResponse = await Amplify.API.query(request: getRequest).response;
  final responseData = getResponse.data;
  if (responseData == null) {
    // throw Exception('GraphQL error: ${getResponse.errors.toString()}');
    safePrint('GraphQL error: ${getResponse.errors.toString()}');
  }
  return responseData;
}

Future<void> deletePlantPlantGroupWithVersion({
  required String responseData,
}) async {
  // Update the data map from response which has _version. We will pass mutated map to mutation request.
  Map<String, dynamic> plantPlantGroupData =
      json.decode(responseData)['getPlantPlantGroup'];
  safePrint('plantPlantGroupData: ${plantPlantGroupData.toString()}');
  // Generate an update request, but modify the input variables to include the _version from previous step.
  final req =
      ModelMutations.delete(PlantPlantGroup.fromJson(plantPlantGroupData));
  final newVars = <String, dynamic>{'input': plantPlantGroupData};
  final reqWithVersion = GraphQLRequest<PlantPlantGroup>(
      document: req.document,
      decodePath: req.decodePath,
      modelType: req.modelType,
      variables: newVars);
  final response = await Amplify.API.mutate(request: reqWithVersion).response;
  final deletedPlantPlantGroup = response.data;
  safePrint('Delete PlantPlantGroup response: ${response.toString()}');
  if (deletedPlantPlantGroup == null) {
    safePrint('Delete PlantPlantGroup GraphQL errors: ${response.errors}');
    return;
  }
  safePrint(
      'Mutation result when deleting PlantPlantGroup: ${deletedPlantPlantGroup.toString()}');
}

Future<void> deletePlantPlantGroupUsingGraphQlString(
    {required String modelId}) async {
  String? responseData =
      await getPlantPlantGroupWithVersionNumber(modelId: modelId);

  if (responseData != null) {
    deletePlantPlantGroupWithVersion(responseData: responseData);
  } else {
    safePrint('Not able to delete PlantPlantGroup object');
  }
}
```

And I basically just am calling this:

```
deletePlantPlantGroupUsingGraphQlString(modelId: '123s332-23lk3-sdlkje-saflkja');
```

And I get the following error resonse:

```
GraphQL error: [GraphQLResponseError{
  "message": "Unable to parse GraphQL query.",
  "errorType": "MalformedHttpRequestException"
}]
```

**Comment by adplant** [comment]
@ragingsquirrel3 Ok, I'm now able to retrieve the object by changing around my query to look like this...

```
Future<String?> getPlantPlantGroupWithVersionNumber(
    {required String modelId}) async {
  const getPlantPlantGroup = 'getPlantPlantGroup';
  // request a model with the _version attached
  String gqlQueryWithVersion = '''query GetPlantPlantGroup(\$id: ID!) {
  getPlantPlantGroup(id: \$id) {
    id
    _version
    plantGroupId
    plantId
  }
}''';
  final getRequest = GraphQLRequest<String>(
    document: gqlQueryWithVersion,
    variables: <String, String>{'id': modelId},
  );
  final getResponse = await Amplify.API.query(request: getRequest).response;
  final responseData = getResponse.data;
  if (responseData == null) {
    // throw Exception('GraphQL error: ${getResponse.errors.toString()}');
    safePrint('GraphQL error: ${getResponse.errors.toString()}');
  }
  return responseData;
}
```

However, now the `deletePlantPlantGroupWithVersion(responseData: responseData)` part fails with this error:

```
Delete PlantPlantGroup response: GraphQLResponse<PlantPlantGroup$> error: [
  {
    "message": "The variables input contains a field that is not defined for input object type 'DeletePlantPlantGroupInput' "
  }
]
Delete PlantPlantGroup GraphQL errors: [GraphQLResponseError{
  "message": "The variables input contains a field that is not defined for input object type 'DeletePlantPlantGroupInput' "
}]
```

**Comment by tauflife** [comment]
@ragingsquirrel3 Any more input. Really need this to start working.

**Comment by adplant** [comment]
@ragingsquirrel3 Any idea on how to resolve this?

**Comment by adplant** [comment]
@ragingsquirrel3 I figured it out. Just so it's documented, I had to change the query to:

```
 String gqlQueryWithVersion = '''query GetPlantPlantGroup(\$id: ID!) {
  getPlantPlantGroup(id: \$id) {
    id
    _version
  }
}''';
```

The fact that I had the `plantGroupId` and `plantId` included was making the `input` fields error.

Of note, we will need to do these manual queries for other queries throughout our app becasue even though the record is deleted from dynamo db, a graphql query will still return that record (it's just that the `_deleted` field will now be `true`).

Getting this sorted out in the future would be very helpful. Couldn't it just be that the `_version`, `_deleted`, and maybe `_createdAt` and `_updatedAt` fields get added to the response?
